In [2]:
import pandas as pd

In [3]:
pd.set_option('display.max_columns', 100)

# 01 — Data Inspection

**Dataset:** ICPSR Study 02538 — Uniform Crime Reporting (UCR), 1960–1962
**Source:** National Archive of Criminal Justice Data (NACJD)
**Format:** Stata `.dta`, one file per year
**Purpose of this notebook:** Read-only audit of the raw data. No cleaning, no transformation.

In [4]:
# importing DataSets
"""
1960 - df60
1961 - df61
1962 - df62
"""
df60 = pd.read_stata('../Data/Raw/02538-0002-Data1960.dta')
df61 = pd.read_stata('../Data/Raw/02538-0004-Data1961.dta')
df62 = pd.read_stata('../Data/Raw/02538-0006-Data1962.dta')

In [5]:
# Shapes of the DataFrame
print(df60.shape)
print(df61.shape)
print(df62.shape)

(20560, 72)
(23232, 72)
(24696, 72)


## Load Confirmation

- 1960: shape (20560, 72)
- 1961: shape (23232, 72)
- 1962: shape (24696, 72)
- All three share the same 72 columns in the same order
- Row counts grow across years — reporting coverage expanded

In [6]:
print("1960"); print(df60.dtypes.value_counts())
print()
print("1961"); print(df61.dtypes.value_counts())
print()
print("1962"); print(df62.dtypes.value_counts())

1960
int32       60
str          3
int16        2
float64      1
category     1
category     1
category     1
category     1
category     1
category     1
Name: count, dtype: int64

1961
int32       60
str          3
int16        2
float64      1
category     1
category     1
category     1
category     1
category     1
category     1
Name: count, dtype: int64

1962
int32       60
str          3
int16        2
float64      1
category     1
category     1
category     1
category     1
category     1
category     1
Name: count, dtype: int64


In [7]:
df60.head()

,CASEID,ST,ORI,GRP,DIV,YEAR,MSA,CD1,CD2,CD3,ADJD,OFF,M1,M2,M3,M4,M5,M6,M7,M8,M9,M10,M11,M12,M13,M14,M15,M16,M17,M18,M19,M20,M21,M22,M23,M24,F1,F2,F3,F4,F5,F6,F7,F8,F9,F10,F11,F12,F13,F14,F15,F16,F17,F18,F19,F20,F21,F22,F23,F24,JW,JB,JI,JO1,JH,JN,AW,AB,AI,AO1,AH,AN
0,NaN,Alabama,AL00103,5,East South Central,60,101,No Adult Male Reported,No Adult Female Reported,No Juvenile Reported,"Age, Race Reported",050,0,0,0,0,0,0,0,0,0,0,1,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,1,0,0,0,0
1,NaN,Alabama,AL00103,5,East South Central,60,101,No Adult Male Reported,No Adult Female Reported,No Juvenile Reported,"Age, Race Reported",060,0,0,0,0,0,1,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,1,0,0,0,0
2,NaN,Alabama,AL00103,5,East South Central,60,101,No Adult Male Reported,No Adult Female Reported,No Juvenile Reported,"Age, Race Reported",150,0,0,0,0,1,1,0,0,0,0,0,0,0,0,0,1,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,3,0,0,0,0,0
3,NaN,Alabama,AL00103,5,East South Central,60,101,No Adult Male Reported,No Adult Female Reported,No Juvenile Reported,"Age, Race Reported",210,0,0,0,0,1,0,0,0,0,0,1,0,0,0,6,4,2,2,2,9,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,1,0,1,0,0,0,0,0,0,0,0,0,0,0,29,2,0,0,0,0
4,NaN,Alabama,AL00103,5,East South Central,60,101,No Adult Male Reported,No Adult Female Reported,No Juvenile Reported,"Age, Race Reported",230,0,0,0,1,2,0,0,0,0,0,0,2,0,1,7,5,9,13,3,23,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,1,1,1,1,1,1,0,0,0,0,0,0,0,0,0,0,63,9,0,0,0,0


## According to the codebook

### -- Linkage Variables --
- **ST** → Numeric State Code (1–2)
- **ORI** → ORI Code. Originating Agency Identifier (3–9)
- **GRP** → Group. Group 0 is Possessions; 1–7 are cities; 8–9 are counties. Sub-Group (Position 11) is blank when not used. All populations are inclusive. (10–11)
- **DIV** → Division. Geographic division in which the state is located (from 1 thru 9). Possessions are coded "0." (12)

### -- Other Agency Information --
- **YEAR** → Year. Last two digits of the year the data reflects, e.g., "85" = 1985, "90" = 1990, etc. (13–14)
- **MSA** → MSA. Metropolitan Statistical Area (MSA) number in which the city is located, if any. Blank if not used. (15–17)
- **SUB** → Suburban. A "suburban" agency is an MSA city with less than 50,000 population (groups 4–7) together with MBA counties (group 9). 0 = Non-Suburban, 1 = Suburban. (18)
- **JUV** → Report Indication. 0 = Juvenile and Adult Reported, 1 = Juvenile Only Reported, 2 = Adult Only Reported, 3 = Not Reported. (19–19)
- **ADJ** → Adjustment. 1960–79: 0 = Age, Race Reported; 1 = No Race Reported; 2 = No Age Reported. 1980–95: 0 = Age, Race, Ethnic origin Reported; 1 = No Age Reported; 2 = No Race Reported; 3 = No Ethnic origin Reported; 4 = No Race or Ethnic origin Reported; 5 = No Age or Ethnic origin Reported; 6 = No Age or Race Reported. (Other codes should be treated as missing or blank.) (20–20)
- **OFF** → Offense. If value is "000," it is the "header" record currently being defined, otherwise it is the "detail" record. (21–23)
- **SEQ** → Sequence Number. A five-digit number which places all cities in alphabetic order, regardless of state. This field is blank for groups 0, 8, and 9. (24–28)
- **CNTY** → County. Three-digit numeric code for the county in which the agency is located. (29–31)
- **CORE** → Core City. Y = agency is core city of an MSA, N = agency is not a core city of an MSA. (Not fully developed until 1974. Use only for 1974 or later.) (32–32)
- **POP** → Current Population. Total Population for the agency for the year reported. (33–42)
- **AGYCT** → Agency Count. Used to accumulate "agencies used" totals in various tabulations. This field is normally "1" but will be "0" for the U.S. Park Police and all state police agencies whose ORI code ends in "SP" (or "99" in California). (Sometimes A or { in early years, meaning is unclear). (43–43)
- **AGY** → Agency Name. (44–67)
- **STATE** → State Name. (68–73)
- **NMO** → Number of Months Reported. (Missing or blank for 1960–1973. Will vary for 1974–1995 and beyond.) (74–75)

### -- Offense Code (OFF) --
- **000** → Header record
- **011** → Murder and Non-Negligent Manslaughter
- **012** → Manslaughter by Negligence
- **020** → Forcible Rape
- **030** → Robbery
- **040** → Aggravated Assault
- **050** → Burglary – Breaking or Entering
- **060** → Larceny – Theft (except motor vehicle)
- **070** → Motor Vehicle Theft
- **080** → Other Assaults
- **090** → Arson
- **100** → Forgery and Counterfeiting
- **110** → Fraud
- **120** → Embezzlement
- **130** → Stolen Property – Buying, Receiving, Poss.
- **140** → Vandalism
- **150** → Weapons – Carrying, Possessing, etc.
- **160** → Prostitution and Commercialized Vice
- **170** → Sex Offenses (except forcible rape and prostitution)
- **180** → Drug Abuse Violations (Total)
  - **181** → Opium and Cocaine and their derivatives (Morphine, Heroin)
  - **182** → Marijuana
  - **183** → Synthetic Narcotics – Manufactured Narcotics which can cause true drug addiction (Demerol, Methadones)
  - **184** → Other Dangerous Non-Narcotic Drugs (Barbiturates, Benzedrine)
  - **185** → Possession (Subtotal)
  - **186** → same as 181
  - **187** → same as 182
  - **188** → Same as 183
  - **189** → same as 184
- **190** → Gambling (Total)
  - **191** → Bookmaking (Horse and Sport Book)
  - **192** → Number and Lottery
  - **193** → All Other Gambling
- **200** → Offenses Against Family and Children
- **210** → Driving Under the Influence
- **220** → Liquor laws
- **230** → Drunkenness
- **240** → Disorderly Conduct
- **250** → Vagrancy
- **260** → All Other Offenses (except traffic)
- **270** → Suspicion
- **280** → Curfew and loitering law Violations
- **290** → Runaways

### -- Male Totals by Age --
- **M1** → Under 11
- **M2** → 11–12
- **M3** → 13–14
- **M4** → 15
- **M5** → 16
- **M6** → 17
- **M7** → No age (Juv)
- **M8** → 18
- **M9** → 19
- **M10** → 20
- **M11** → 21
- **M12** → 22
- **M13** → 23
- **M14** → 24
- **M15** → 25–29
- **M16** → 30–34
- **M17** → 35–39
- **M18** → 40–44
- **M19** → 45–49
- **M20** → 50–54
- **M21** → 55–59
- **M22** → 60–64
- **M23** → Over 64
- **M24** → No Age (Adult)

### -- Female Totals by Age --
- **F1** → Under 11
- **F2** → 11–12
- **F3** → 13–14
- **F4** → 15
- **F5** → 16
- **F6** → 17
- **F7** → No age (Juv)
- **F8** → 18
- **F9** → 19
- **F10** → 20
- **F11** → 21
- **F12** → 22
- **F13** → 23
- **F14** → 24
- **F15** → 25–29
- **F16** → 30–34
- **F17** → 35–39
- **F18** → 40–44
- **F19** → 45–49
- **F20** → 50–54
- **F21** → 55–59
- **F22** → 60–64
- **F23** → Over 64
- **F24** → No Age (Adult)

### -- Juvenile Totals by Race and Ethnic Origin --
- **JW** → Juvenile-White
- **JB** → Juvenile-Negro (Black)
- **JI** → Juvenile-Indian
- **JO1** → Juvenile-Chinese
- **JH** → Juvenile-Japanese
- **JN** → Juvenile-Other

### -- Adult Totals by Race and Ethnic Origin --
- **AW** → Adult-White
- **AB** → Adult-Negro (Black)
- **AI** → Adult-Indian
- **AO1** → Adult-Chinese
- **AH** → Adult-Japanese
- **AN** → Adult-Other

---



## Missing Values

Null counts per column per year. A "column with nulls" is not automatically a problem — some columns (like `MSA`, `NMO`) are documented as blank under certain conditions. The goal here is to catalogue, not to judge yet.

In [8]:
for name, df in [("1960", df60), ("1961", df61), ("1962", df62)]:
    nulls = df.isnull().sum()
    nulls = nulls[nulls > 0].sort_values(ascending=False)
    print(f"\n=== {name} — columns with nulls ===")
    print(nulls if len(nulls) else "None")


=== 1960 — columns with nulls ===
CASEID    20560
dtype: int64

=== 1961 — columns with nulls ===
CASEID    23232
dtype: int64

=== 1962 — columns with nulls ===
CASEID    24696
dtype: int64


In [9]:
# So We now know, while cleaning the data we will have to drop "CASEID" as it is null for all rows

In [10]:
# Checking Dublicate Rows:
for name, df in [("1960", df60), ("1961", df61), ("1962", df62)]:
    print(f"{name}: full-row duplicates = {df.duplicated().sum()}")

1960: full-row duplicates = 2
1961: full-row duplicates = 0
1962: full-row duplicates = 0


In [11]:
KEY = ['ORI', 'YEAR', 'OFF']
for name, df in [("1960", df60), ("1961", df61), ("1962", df62)]:
    dupes = df.duplicated(subset=KEY).sum()
    print(f"{name}: duplicates on {KEY} = {dupes}")

1960: duplicates on ['ORI', 'YEAR', 'OFF'] = 2
1961: duplicates on ['ORI', 'YEAR', 'OFF'] = 0
1962: duplicates on ['ORI', 'YEAR', 'OFF'] = 0


In [12]:
#There is only 2 dublicate pairs in df60 and none in df61 and df62

In [13]:
dupes_60 = df60[df60.duplicated(subset=['ORI', 'YEAR', 'OFF'], keep=False)]

# Show just the identifying columns and a few counts
dupes_60[['ST', 'ORI', 'GRP', 'DIV', 'OFF']].sort_values(['ORI', 'OFF'])

,ST,ORI,GRP,DIV,OFF
10898,Missouri,MO09581,6,West North Central,110
10899,Missouri,MO09581,6,West North Central,110
20091,Wisconsin,WIMPD00,1B,East North Central,050
20092,Wisconsin,WIMPD00,1B,East North Central,050


In [14]:
# Lets check the Whole column and compare the Values before  making a decision

In [15]:
# first dublicate 
df60.iloc[[10898, 10899],:]

,CASEID,ST,ORI,GRP,DIV,YEAR,MSA,CD1,CD2,CD3,ADJD,OFF,M1,M2,M3,M4,M5,M6,M7,M8,M9,M10,M11,M12,M13,M14,M15,M16,M17,M18,M19,M20,M21,M22,M23,M24,F1,F2,F3,F4,F5,F6,F7,F8,F9,F10,F11,F12,F13,F14,F15,F16,F17,F18,F19,F20,F21,F22,F23,F24,JW,JB,JI,JO1,JH,JN,AW,AB,AI,AO1,AH,AN
10898,NaN,Missouri,MO09581,6,West North Central,60,754,No Adult Male Reported,No Adult Female Reported,No Juvenile Reported,"Age, Race Reported",110,0,0,0,0,1,0,0,0,0,0,0,0,0,0,1,0,1,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,3,0,0,0,0,0
10899,NaN,Missouri,MO09581,6,West North Central,60,754,No Adult Male Reported,No Adult Female Reported,No Juvenile Reported,"Age, Race Reported",110,0,0,0,0,1,0,0,0,0,0,0,0,0,0,1,0,1,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,3,0,0,0,0,0


In [16]:
# Second Dublicate
df60.iloc[[20091,20092],:]

,CASEID,ST,ORI,GRP,DIV,YEAR,MSA,CD1,CD2,CD3,ADJD,OFF,M1,M2,M3,M4,M5,M6,M7,M8,M9,M10,M11,M12,M13,M14,M15,M16,M17,M18,M19,M20,M21,M22,M23,M24,F1,F2,F3,F4,F5,F6,F7,F8,F9,F10,F11,F12,F13,F14,F15,F16,F17,F18,F19,F20,F21,F22,F23,F24,JW,JB,JI,JO1,JH,JN,AW,AB,AI,AO1,AH,AN
20091,NaN,Wisconsin,WIMPD00,1B,East North Central,60,538,No Adult Male Reported,No Adult Female Reported,No Juvenile Reported,"Age, Race Reported",050,0,0,363,108,112,98,0,26,20,15,20,15,23,20,40,24,12,5,3,4,0,0,0,0,0,0,7,7,4,3,0,1,0,1,1,1,0,0,1,0,0,0,1,0,0,0,0,0,0,0,0,0,0,0,567,356,10,0,0,2
20092,NaN,Wisconsin,WIMPD00,1B,East North Central,60,538,No Adult Male Reported,No Adult Female Reported,No Juvenile Reported,"Age, Race Reported",050,0,0,363,108,112,98,0,26,20,15,20,15,23,20,40,24,12,5,3,4,0,0,0,0,0,0,7,7,4,3,0,1,0,1,1,1,0,0,1,0,0,0,1,0,0,0,0,0,0,0,0,0,0,0,567,356,10,0,0,2


### By the Above inspection we can say that both the pair of rows are dublicate.

## Now we will check if any row has ha value that does not match the codebook

In [27]:
for name, df in [("1960", df60), ("1961", df61), ("1962", df62)]:
    print(f"\n{name}")
    for col in ['DIV', 'GRP', 'OFF']:
        print(f"\n{col}")
        print(df[col].value_counts().head(15))


1960

DIV
DIV
East North Central    4797
Middle Atlantic       4788
Pacific               3441
New England           2043
South Atlantic        1549
West North Central    1366
West South Central    1195
Mountain               827
East South Central     524
Possessions             30
Name: count, dtype: int64

GRP
GRP
5     5299
6     4273
4     3116
3     3116
2     1305
9B    1303
1C     655
9A     440
1B     396
9C     277
7      261
1A     119
Name: count, dtype: int64

OFF
OFF
060    1231
240    1214
050    1210
210    1192
230    1192
260    1134
070    1094
080    1088
170     970
150     862
250     834
040     833
220     829
200     814
110     758
Name: count, dtype: int64

1961

DIV
DIV
East North Central    5455
Middle Atlantic       5073
Pacific               3963
New England           2458
South Atlantic        1706
West North Central    1685
West South Central    1435
Mountain               828
East South Central     629
Name: count, dtype: int64

GRP
GRP
6     5772
5  

#### Everything follows as per codebook

In [26]:
for name, df in [("1960", df60), ("1961", df61), ("1962", df62)]:
    print(f"{name}: OFF-->'000' rows = {(df['OFF'] == '000').sum()}")

1960: OFF-->'000' rows = 0
1961: OFF-->'000' rows = 0
1962: OFF-->'000' rows = 0


In [29]:
# Quick pass over each year:
#   - any negative counts (would be impossible in arrest data)
#   - columns that are the same value all the way down (dead weight)
#   - numeric columns that are all zeros (either unused or never reported)

for name, df in [("1960", df60), ("1961", df61), ("1962", df62)]:
    numeric = df.select_dtypes('number')
    print(f"\n----> {name} <-----")

    negs = (numeric < 0).sum().sum()
    print(f"negatives--> {negs}")

    counts = df.nunique()
    dead = counts[counts <= 1].index.tolist()
    print(f"constant cols--> {dead}")

    zeros = [c for c in numeric.columns if (numeric[c] == 0).all()]
    print(f"all-zero cols--> {zeros}")


----> 1960 <-----
negatives--> 0
constant cols--> ['CASEID', 'YEAR', 'CD1', 'CD2', 'CD3', 'ADJD', 'M1', 'M2', 'M7', 'M21', 'M22', 'M23', 'F1', 'F2', 'F7', 'F21', 'F22', 'F23', 'JW', 'JB', 'JI', 'JO1', 'JH', 'JN']
all-zero cols--> ['M1', 'M2', 'M7', 'M21', 'M22', 'M23', 'F1', 'F2', 'F7', 'F21', 'F22', 'F23', 'JW', 'JB', 'JI', 'JO1', 'JH', 'JN']

----> 1961 <-----
negatives--> 0
constant cols--> ['CASEID', 'YEAR', 'M1', 'M2', 'M7', 'M21', 'M22', 'M23', 'F1', 'F2', 'F7', 'F21', 'F22', 'F23', 'JW', 'JB', 'JI', 'JO1', 'JH', 'JN']
all-zero cols--> ['M1', 'M2', 'M7', 'M21', 'M22', 'M23', 'F1', 'F2', 'F7', 'F21', 'F22', 'F23', 'JW', 'JB', 'JI', 'JO1', 'JH', 'JN']

----> 1962 <-----
negatives--> 0
constant cols--> ['CASEID', 'YEAR', 'CD1', 'CD2', 'CD3', 'ADJD', 'M1', 'M2', 'M7', 'M21', 'M22', 'M23', 'F1', 'F2', 'F7', 'F21', 'F22', 'F23', 'JW', 'JB', 'JI', 'JO1', 'JH', 'JN']
all-zero cols--> ['M1', 'M2', 'M7', 'M21', 'M22', 'M23', 'F1', 'F2', 'F7', 'F21', 'F22', 'F23', 'JW', 'JB', 'JI', 'JO1', 

The following columns are 0 for every row in 1960-1962:

**Age extremes (male):** M1 (under 11), M2 (11–12), M7 (no age, juv),
M21 (55–59), M22 (60–64), M23 (over 64)

**Age extremes (female):** F1, F2, F7, F21, F22, F23

**All juvenile-by-race:** JW, JB, JI, JO1, JH, JN

The following columns are constant all over in year 1960-1962

**Case Type Reported:** CD1(No Adult Male Reported), CD2(No Adult Female Reported), CD3(No Juvenile Reported), ADJD(Age, Race Reported)

#### We did not include YEAR as it should be constant, CASEID because it is NaN for all rows and everyother is already counted in 0s columns

### From this we know what more columns to drop
**Till now:** 'M1', 'M2', 'M7', 'M21', 'M22', 'M23', 'F1', 'F2', 'F7', 'F21', 'F22', 'F23', 'JW', 'JB', 'JI', 'JO1', 'JH', 'JN', 'CASEID' and optionals including 'CD1','CD2','CD3','ADJD'
**ROWS:** Only in df60 or 1960 - '10899', '20092'

# So, Here ends our inspections, We studied what things can be removed and is unnecessary. Data Structure and types. Columns and its meaning and Codebook. Now we move towards cleaning the data!!